source/common/unified_memory_allocation/src/memory_tracker.cpp

```diff
@@ -0,0 +1,170 @@
+        std::unique_lock<std::shared_mutex> lock(mtx);
+
+        map.erase(reinterpret_cast<uintptr_t>(ptr));
+
+        // TODO: handle removing part of the range
+        (void)size;
+
+        return UMA_RESULT_SUCCESS;
+    }
+
+    void *find(const void *ptr) {
+        std::shared_lock<std::shared_mutex> lock(mtx);
+
+        auto intptr = reinterpret_cast<uintptr_t>(ptr);
+        auto it = map.upper_bound(intptr);
+        if (it == map.begin()) {
+            return nullptr;
+        }
+
+        --it;
+        if (intptr >= it->first && intptr < it->first + it->second.first) {
+            return it->second.second;
+        }
+
+        return nullptr;
+    }
+
+  private:
+    std::shared_mutex mtx;
+    std::map<uintptr_t, std::pair<size_t, void *>> map;
+};
+
+static enum uma_result_t
+umaMemoryTrackerAdd(uma_memory_tracker_handle_t hTracker, void *pool,
+                    const void *ptr, size_t size) {
+    return hTracker->add(pool, ptr, size);
+}
+
+static enum uma_result_t
+umaMemoryTrackerRemove(uma_memory_tracker_handle_t hTracker, const void *ptr,
+                       size_t size) {
+    return hTracker->remove(ptr, size);
+}
+
+extern "C" {
+
+uma_memory_tracker_handle_t umaMemoryTrackerGet() {
+    static uma_memory_tracker_t tracker;
+    return &tracker;
+}
+
+void *umaMemoryTrackerGetPool(uma_memory_tracker_handle_t hTracker,
+                              const void *ptr) {
+    return hTracker->find(ptr);
+}
+
+struct uma_tracking_memory_provider_t {
+    uma_memory_provider_handle_t hUpstream;
+    uma_memory_tracker_handle_t hTracker;
+    uma_memory_pool_handle_t pool;
+};
+
+typedef struct uma_tracking_memory_provider_t uma_tracking_memory_provider_t;
+
+static enum uma_result_t trackingAlloc(void *hProvider, size_t size,
+                                       size_t alignment, void **ptr) {
+    uma_tracking_memory_provider_t *p =
+        (uma_tracking_memory_provider_t *)hProvider;
+    enum uma_result_t ret = UMA_RESULT_SUCCESS;
+
+    ret = umaMemoryProviderAlloc(p->hUpstream, size, alignment, ptr);
+    if (ret != UMA_RESULT_SUCCESS) {
+        return ret;
+    }
+
+    // TODO: is it fine that allocation and adding memory to tracker is not a single atomic action
+    ret = umaMemoryTrackerAdd(p->hTracker, p->pool, *ptr, size);
+    if (ret != UMA_RESULT_SUCCESS && p->hUpstream) {
+        if (umaMemoryTrackerRemove(p->hTracker, *ptr, size)) {
+            // TODO: LOG
+        }
+    }
+
+    return ret;
+}
+
+static enum uma_result_t trackingFree(void *hProvider, void *ptr, size_t size) {
+    enum uma_result_t ret;
+    uma_tracking_memory_provider_t *p =
+        (uma_tracking_memory_provider_t *)hProvider;
+
+    ret = umaMemoryTrackerRemove(p->hTracker, ptr, size);
+    if (ret != UMA_RESULT_SUCCESS) {
+        return ret;
+    }
+
+    // TODO: is it fine that deallocation and removing memory from the tracker is not a single atomic action
```

Comment:
yes, because nothing else should have any pointer in this specific memory range. that would be race at a higher level.